nucleof072rb/Core/Src/main.c

```diff
@@ -1,181 +1,248 @@
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+#define MAX_VAL 1023
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+const float MAX_DUTY_CYCLE = .1;
+const float MIN_DUTY_CYCLE = .05;
+const int INPUT_CLOCK = 48e6; //48 MHz
+const int PRESCALER = 14;
+const int FREQUENCY = 50;
+const int COUNTER_PERIOD = (INPUT_CLOCK/(PRESCALER+1))/FREQUENCY;
+const int MAX_VALUE = 1023;
+
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+HAL_StatusTypeDef read_adc(uint16_t * data);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+//  SPI_HandleTypeDef hspi;
+//  HAL_SPI_MspInit(&hspi);
+
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+  // Bring CS line (PB8) high initially
+  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+  // Start timer
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+	  // Read and parse data
+	  uint16_t data;
+	  HAL_StatusTypeDef success = read_adc(&data);
+
+	  if (success == HAL_OK) {
+		  uint16_t value = (COUNTER_PERIOD * MAX_DUTY_CYCLE - COUNTER_PERIOD * MIN_DUTY_CYCLE)
+			  * data/MAX_VALUE;

```

Comment:
make this calculation clearer and more efficient. use _HAL_TIM_GET_AUTORELOAD to get the counter period